Bugzilla/DuoClient.pm

```diff
@@ -0,0 +1,282 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+# This Source Code Form is "Incompatible With Secondary Licenses", as
+# defined by the Mozilla Public License, v. 2.0.
+#
+# Adapted from https://github.com/duosecurity/duo_universal_python/blob/main/duo_universal/client.py
+
+package Bugzilla::DuoClient;
+
+use 5.10.1;
+use Moo;
+
+use Bugzilla::Constants;
+use Bugzilla::Error;
+use Bugzilla::Logging;
+use Bugzilla::Util qw(generate_random_password mojo_user_agent);
+
+use Mojo::JWT;
+use Mojo::URL;
+use Try::Tiny;
+use Types::Standard -types;
+
+use constant JTI_LENGTH              => 36;
+use constant MINIMUM_STATE_LENGTH    => 22;
+use constant MAXIMUM_STATE_LENGTH    => 1024;
+use constant FIVE_MINUTES_IN_SECONDS => 300;
+
+use constant ERR_USERNAME     => 'The Duo username was invalid.';
+use constant ERR_CODE         => 'The Duo authorization code was missing.';
+use constant ERR_HEALTH_CHECK => 'The Duo service health check failed.';
+use constant ERR_TOKEN_ERROR  => 'The Duo service had an error obtaining authorization token.';
+use constant ERR_STATE_LEN => 'The Duo state must be at least '
+  . MINIMUM_STATE_LENGTH
+  . ' characters long and no longer than '
+  . MAXIMUM_STATE_LENGTH
+  . ' characters.';
+
+use constant OAUTH_V1_HEALTH_CHECK_ENDPOINT => '/oauth/v1/health_check';
+use constant OAUTH_V1_AUTHORIZE_ENDPOINT    => '/oauth/v1/authorize';
+use constant OAUTH_V1_TOKEN_ENDPOINT        => '/oauth/v1/token';
+use constant CLIENT_ASSERT_TYPE =>
+  'urn:ietf:params:oauth:client-assertion-type:jwt-bearer';
+
+#########################
+#    Initialization     #
+#########################
+
+has host          => (is => 'ro', required => 1, isa => Str);
+has client_id     => (is => 'ro', required => 1, isa => Str);
+has client_secret => (is => 'ro', required => 1, isa => Str);
+has redirect_uri => (
+  is      => 'ro',
+  default => Bugzilla->localconfig->urlbase . 'mfa/duo/callback'
+);
+
+#########################
+#    Private Methods    #
+#########################
+
+# Standard JWT args to include with all JWT tokens
+sub _create_jwt_args {
+  my ($self, $endpoint) = @_;
+
+  my $jwt_args = {
+    'iss' => $self->client_id,
+    'sub' => $self->client_id,
+    'aud' => $endpoint,
+    'exp' => time() + FIVE_MINUTES_IN_SECONDS,
+    'jti' => generate_random_password(JTI_LENGTH)
+  };
+
+  return $jwt_args;
+}
+
+# Help for creating a URL object with default parameters
+sub _create_api_uri {
+  my ($self, $path) = @_;
+
+  my $uri = Mojo::URL->new;
+  $uri->host($self->host);
+  $uri->path($path);
+
+  # If we are running tests, then use http
+  if ($self->host =~ /(externalapi|localhost)/) {
+    $uri->scheme('http');
+  }
+  else {
+    $uri->scheme('https');
+  }
```

Comment:
I've noticed we are checking for `localhost` or `externalapi` as the hostname in various places and making changes to parameters based on that. Can we somehow remove this "test environment only" condition and instead have parameters that we can change on a per-environment basis?

Reply:
This is because when I run the docker cluster locally to use BMO myself on my workstation, the duo host needs to be different if we are redirecting the user to the Duo (mocked) site or if the DuoClient code is calling Duo on the backend to exchange the auth code for a token. In production, the external host will be the same in both cases.  I can do what you suggest and pass in an environment variable when running under the test environment and look for that instead of matching on hostname. Or if you think of a better way.